feat/efc: add API for read/erase/write flash

tmplt opened the pull request.

martinmortsell committed:
Add Flash Sector Tokens and read_word Funcitonality (WIP)

Add Erase Sector Method

Add Write Page Method

Add Write Word Method

Add Two Bootloader Examples

tmplt commented:
CC @martinmortsell

martinmortsell committed:
Refractor Flash API, Move Sectors Struct Into Efc Struct

Update `efc.rs` Example

Add Flash Sector Tokens and read_word Funcitonality (WIP)

Add Erase Sector Method

Add Write Page Method

Add Write Word Method

Refractor Flash API, Move Sectors Struct Into Efc Struct

Update `efc.rs` Example

Fix Issues Causing `docs` Generation To Fail

Fix Clippy Lints

Fix Issues With `atsamv71_xult` Examples

Run `cargo fmt`

Run `cargo fmt` On Examples

martinmortsell commented:
All checks are finally passing.

Currently a user could generate multiple tokens for the same flash sector, I'm going to look into making sure that only one set of `Sectors` can be generated using the `EFC` singleton from the `pac`.

I am uncertain about the `write_page` method, in the past I've seen the compiler optimize away writes, but I have not managed to trigger that behaviour this time.

Apart from that, it would be nice to have better support for writing across multiple sectors, e.g. iterating across a subset of the sectors or similar.

martinmortsell committed:
Implement `embedded-storage` Flash API

Remove Dead Code

Conform To Clippy And Rusfmt

Update Examples

Update `CHANGELOG.md`

Run `fmt` On Examples

martinmortsell committed:
Improve Readability of Write And Erase Methods

Add Doc Comment To CAPACITY const

martinmortsell commented:
I have basically rewritten the FLash API from scratch, as I found the `embedded-storage` crate which defines traits for flash transfers.
I'll look through the code tomorrow and see if I can break it somehow, but this is mostly ready for review now.

tmplt commented:
I suggest rebasing upon master, and some squashing, to simplify the review.

martinmortsell committed:
`fmt`

martinmortsell committed:
Revert Changes Made To Examples Unrelated To PR

martinmortsell committed:
Improve Top Level Efc Documentation

martinmortsell committed:
Fix efc Doc Example
